### PR TITLE
Add support of resource range binding

### PR DESCRIPTION
**Pull request by jiaolu**
If there is user data descriptors range[array] declaration in api, and shader resource
binding in the shader can be set as element of range array.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1b2557c9f69d58766eee6c1484b3f87a201d5065
<details><summary>CTS tests (Failed: 0/186371)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 37104/65208 (56.9%)</li>
<li>Failed: 0/65208 (0.0%)</li>
<li>Not Supported: 28104/65208 (43.1%)</li><li>Warnings: 0/65208 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30377/55953 (54.3%)</li>
<li>Failed: 0/55953 (0.0%)</li>
<li>Not Supported: 25576/55953 (45.7%)</li><li>Warnings: 0/55953 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39403/65210 (60.4%)</li>
<li>Failed: 0/65210 (0.0%)</li>
<li>Not Supported: 25807/65210 (39.6%)</li><li>Warnings: 0/65210 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 857c4b77ba6b38af0440432d349e46ec47b8e883
<details><summary>CTS tests (Failed: 0/186371)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 37104/65208 (56.9%)</li>
<li>Failed: 0/65208 (0.0%)</li>
<li>Not Supported: 28104/65208 (43.1%)</li><li>Warnings: 0/65208 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30377/55953 (54.3%)</li>
<li>Failed: 0/55953 (0.0%)</li>
<li>Not Supported: 25576/55953 (45.7%)</li><li>Warnings: 0/55953 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39403/65210 (60.4%)</li>
<li>Failed: 0/65210 (0.0%)</li>
<li>Not Supported: 25807/65210 (39.6%)</li><li>Warnings: 0/65210 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by jiaolu** [comment]
> The implementation looks go to me.
> 
> Please add tests for this. This is a major new functionality. We need something to make sure it does what we think it should do.

Test added, any more comment?

File: lgc/builder/DescBuilder.cpp [hunk 1]
```diff
@@ -91,6 +96,7 @@ Value *DescBuilder::CreateLoadBufferDesc(unsigned descSet, unsigned binding, Val
         desc = UndefValue::get(descTy);
       } else {
         dwordOffset += node->offsetInDwords;
+        dwordOffset += (binding - node->binding) * node->stride;
```
Comment (s-perron) [inline lgc/builder/DescBuilder.cpp:99]: I don't think is being tested.  If I understand the test correctly, the resource is not in the root table.  This code is executed only if the the resource is in the root table.  I think another test should be added to run all of the possible paths.

File: lgc/builder/DescBuilder.cpp [hunk 2]
```diff
@@ -74,6 +74,11 @@ Value *DescBuilder::CreateLoadBufferDesc(unsigned descSet, unsigned binding, Val
       return UndefValue::get(getBufferDescTy(pointeeTy));
     }
 
+    if (node->binding != binding) {
+      // We found a binding in the middle of a range. Add the difference to the index.
+      descIndex = CreateAdd(descIndex, getInt32(binding - node->binding));
```
Comment (s-perron) [inline lgc/builder/DescBuilder.cpp:79]: I don't know LLVM well enough, but will descIndex be a constant if the two inputs to CreateAdd are constants?  If not then the next if condition will always be false.  You should have a test that tests both the descIndex being constant and a variable.

**Review by s-perron — COMMENTED** [review]
Some extra testing might be in order just to make sure you are testing all of the new code.

File: lgc/builder/DescBuilder.cpp (hunk 2 above)
Comment (s-perron) [reply lgc/builder/DescBuilder.cpp:79]: Line 99 will be executed by other tests, but `binding - node->binding` will always be 0 in those cases.  We need a test where it is not 0.

Comment (s-perron) [reply lgc/builder/DescBuilder.cpp:79]: > It looks like this branch is never hit in the tests

Right, if the descriptor being loaded is not a buffer descriptor, this function will not be called.  The test has a range of Resource descriptors.  To test the code in this function, you need a range of buffer descriptors.

**Comment by amdvlk-admin** [comment]
### Test summary for commit c2e91d8d5ba99e549061b3cf6b074bc7e8b4d5f4
<details><summary>CTS tests (Failed: 0/179976)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35765/63748 (56.1%)</li>
<li>Failed: 0/63748 (0.0%)</li>
<li>Not Supported: 27983/63748 (43.9%)</li><li>Warnings: 0/63748 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 27857/52478 (53.1%)</li>
<li>Failed: 0/52478 (0.0%)</li>
<li>Not Supported: 24621/52478 (46.9%)</li><li>Warnings: 0/52478 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 38010/63750 (59.6%)</li>
<li>Failed: 0/63750 (0.0%)</li>
<li>Not Supported: 25740/63750 (40.4%)</li><li>Warnings: 0/63750 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by jiaolu** [comment]
@s-perron , The test had been updated for a while, any comment?

File: lgc/test/TextureRange.lgc [hunk 3]
```diff
@@ -1,50 +1,75 @@
...
+; CHECK: %{{.*}} = getelementptr i8, i8 addrspace(4)* [[desc1]], i32 32
 
 ; RUN: lgc -mcpu=gfx1030 -o - - <%s | FileCheck --check-prefixes=SHADER_TEST %s
 ; SHADER_TEST: s_load_dwordx8 [[desc:.*]], {{.*}}, 0x20
 ; SHADER_TEST: image_sample {{.*}}, {{.*}}, [[desc]], {{.*}} dmask:0xf
-
+; SHADER_TEST: s_load_dwordx4 {{.*}}, {{.*}}, 0x10
+; SHADER_TEST: .registers:
+; SHADER_TEST: (SPI_SHADER_USER_DATA_PS_4): 0x6
+; SHADER_TEST: (SPI_SHADER_USER_DATA_PS_5): 0x7
```
Comment (s-perron) [inline lgc/test/TextureRange.lgc:16]: I don't see where s4 and s5 are used in the GCN.  I cannot be sure they are being used correctly.

**Comment by amdrexu** [comment]
@jiaolu Please update the PR.

**Comment by jiaolu** [comment]
> @jiaolu Please update the PR.

The PR and checks are updated according reviewer's points.

**Review by amdrexu — APPROVED** [review]

**Comment by kuhar** [comment]
@jiaolu @amdrexu Seems like there's one lgc test failure caused by this PR: https://github.com/GPUOpen-Drivers/llpc/runs/7314487224?check_suite_focus=true

**Pull request merged**